keps/sig-api-machinery/3158-optional-maximum-limit-on-list-requests-to-etcd/README.md

```diff
@@ -0,0 +1,901 @@
+- apiserver_list_duration_p99 is within defined scalability SLO
+
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+A user deployed an application as daemonset that queries pods belonging to the node it is running. It translates to range query etcd from `/registry/pods/` to `/registry/pods0` and each such query payload is close to 2GiB. 
+```
+{
+    "level": "warn",
+    "ts": "2022-03-31T23:00:06.771Z",
+    "caller": "etcdserver/util.go:163",
+    "msg": "apply request took too long",
+    "took": "111.479086ms",
+    "expected-duration": "100ms",
+    "prefix": "read-only range ",
+    "request": "key:\"/registry/pods/\" range_end:\"/registry/pods0\" ",
+    "response": "range_response_count:27893 size:191330794"
+}
+```
+
+```
+I0331 23:00:06.844303      10 trace.go:205] Trace[484169231]: "List etcd3" key:/pods,resourceVersion:,resourceVersionMatch:,limit:0,continue: (31-Mar-2022 23:00:04.575) (total time: 2268ms):
+Trace[484169231]: [2.268397203s] [2.268397203s] END
+
+I0331 23:00:06.848003      10 trace.go:205] Trace[638277422]: "List" url:/api/v1/namespaces//pods,user-agent:OpenAPI-Generator/12.0.1/python,client:10.13.20.49 (31-Mar-2022 23:00:04.575) (total time: 2272ms):
+Trace[638277422]: ---"Listing from storage done" 2268ms (23:00:00.844)
+Trace[638277422]: [2.272115626s] [2.272115626s] END
+```
+10x clients issued such queries lead to etcd OOM-killed and failed other APIServer requests hitting to that node.
+
+With the approach and `max-list-etcd-limit = 500`, user should expect 
+- APIServer is healthy and serving requests because etcd cluster is healthy 
+- List etcd duration increases from 2 seconds around 3 seconds and this duration is available as a prometheus metric exposed in "/metrics" endpoint. 
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional) 
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- Setting the maximum limit too low will increase the total round-trip time between APIServer and etcd to breach the default 1 minute request time. This will cause APIServer cancelles the requests prematurely. Example https://github.com/kubernetes/kubernetes/pull/94303. 
+
+- Use the maximum limit as suggestion and adjust it dynamically to fit in the 1 minute request timeout. For example, assuming each resource object size is identical given resource type, if the very first list with suggested maximum limit takes 2s but requires another 60 such subsequent list, double the limit to reduce the round trip time. 
```

Comment:
Dynamically choosing the page size (that too per resource-type) is not just tedious, but also hard to get right IMO. For e.g these objects can grow over time, or round-trip duration may vary considerably in setups where etcd and apiserver are talking to each other over a network. Also the terminology "adjust it dynamically" isn't clear here - whether it is an operator externally changing the flag values on apiserver or apiserver internally adjusting its values. I'd suggest making these points clear in the caveat.

+cc @wojtek-t - for any thoughts

Reply:
Also another perspective at this would be to push the client itself to paginate its list calls to apiserver, to avoid running into the above situation in the first place. This is also a general k8s guidance.

Reply:
After internal experiment and testing, I don't think dynamically choosing the page size is necessary as this only changes requests in a subset of list calls (almost every well-behaved client requests will be untouched and the behavior is just as it is today)

So, I'd like to remove this block.